source/robot_model/src/Model.cpp

```
@@ -5,277 +5,381 @@
-  for (unsigned int i = 0; i < nb_joints; ++i) {
-    for (unsigned int j = 0; j < nb_joints; ++j) {
-      coefficients.emplace_back(Eigen::Triplet<double>(i, j, hessian_matrix(i, j)));
+
+  StateRepresentation::JointPositions Model::inverse_geometry(StateRepresentation::JointState actual_joint_state,
+                                                              const StateRepresentation::CartesianState &desired_cartesian_state)
+  {
+    // This method uses the angular velocity evaluated by the 'Infeasibility-free inverse kinematics method' which takes into account the
+    // joint limits. The angular velocity is then used for indicating the direction for an iterative algorithm.
+    // To converge at the desired end-effector position a similar approach to the Newton–Raphson method is used, the only difference is that
+    // here the pseudo-inverse of the jacobian is not used and instead is used the angualr velocity evaluated by the
+    // 'Infeasibility-free inverse kinematics method'.
+
+    const int JOINT_ID = this->robot_model_.njoints - 2;
+
+    // TODO if the actual_joint_state is not provided initialize the position of the robot in a radom conformation
+    // TODO take as input the parameters eps and IT_MAX
+    // TODO return 'success'
+
+    Eigen::VectorXd q = actual_joint_state.get_positions(); //pinocchio::neutral(model);
+    const double eps = 1e-4;
+    const int IT_MAX = 1000;
+
+    std::vector<unsigned int> frame_ids{(unsigned int)JOINT_ID};
+
+    StateRepresentation::CartesianPose actual_cartesian_state;
+    double error;
+
+    bool success = false;
+    Eigen::VectorXd v(this->robot_model_.nv);
+    for (int i = 0;; i++)
+    {
+
+      actual_cartesian_state = this->forward_geometry(actual_joint_state, frame_ids).front();
+      error = desired_cartesian_state.dist(actual_cartesian_state, StateRepresentation::CartesianStateVariable::POSE);
+
+      if (error < eps)
+      {
+        success = true;
```

success is never used right? You could throw an exception (that you need to create) at the end in case of non convergence.

---

Agree that the `success` flag is unused and that the `break` conditions are not neat (use a `while` loop for the iterator instead).

Regarding exception in non-convergence, this is a design question to think about. Should the user be able to use a simple conditional `if success, use the result`, or do we force `try inverse geometry, catch exception` on the user side? It probably makes sense to fail hard and have the user deal with it explicitly though

---

no so all those `throw` code in the cpp file. Like this. You can create a `MySpecificException` class in the `Exceptions` folder and call it with a `throw`